mythril/laser/smt/independence_solver.py

```diff
@@ -0,0 +1,147 @@
+import z3
+
+from mythril.laser.smt import Model
+from typing import Set
+
+
+def _get_expr_variables(expression: z3.ExprRef):
+    """
+    Get's the variables that make up the current expression
+    :param expression:
+    :return:
+    """
+    result = []
+    if not expression.children() and not isinstance(expression, z3.BitVecNumRef):
+        result.append(expression)
+    for child in expression.children():
+        c_children = _get_expr_variables(child)
+        result.extend(c_children)
+    return result
+
+
+class DependenceBucket:
+    """ Bucket object to contain a set of conditions that are dependent on each other """
+
+    def __init__(self, variables=None, conditions=None):
+        """
+        Initializes a DependenceBucket object
+        :param variables: Variables contained in the conditions
+        :param conditions: The conditions that are dependent on each other
+        """
+        self.variables = variables or []
+        self.conditions = conditions or []
+
+
+class DependenceMap:
+    """ DependenceMap object that maintains a set of dependence buckets, used to separate independent smt queries """
+
+    def __init__(self):
+        """ Initializes a DependenceMap object """
+        self.buckets = []
+        self.variable_map = {}
+
+    def add_condition(self, condition: z3.BoolRef):
+        """
+        Add condition to the dependence map
+        :param condition: The condition that is to be added to the dependence map
+        """
+        variables = set(_get_expr_variables(condition))
+        relevant_buckets = set()
+        for variable in variables:
+            try:
+                bucket = self.variable_map[str(variable)]
+                relevant_buckets.add(bucket)
+            except KeyError:
+                continue
+
+        new_bucket = DependenceBucket(variables, [condition])
+        self.buckets.append(new_bucket)
+
+        if relevant_buckets:
+            # Merge buckets, and rewrite variable map accordingly
+            relevant_buckets.add(new_bucket)
+            new_bucket = self.merge_buckets(relevant_buckets)
+
+        for variable in variables:
+            self.variable_map[str(variable)] = new_bucket
+
+    def merge_buckets(self, bucket_list: Set[DependenceBucket]):
```

Comment:
It might be good for this function to be hidden right?

Reply:
And also a set named as bucket_list might be a bit misleading

Reply:
Well it was a list at first, and I liked the name so much I didn't change it

Reply:
Is that argument convincing enough to resolve this conversation :innocent: ?

Reply:
and about making the `merge_buckets()` hidden? Would it be intended to be called from outside at any time?It might more readble if it's hidden in that case right?

Reply:
I think I already did that